cpp/oneapi/dal/table/backend/csr_table_impl.hpp

```diff
@@ -0,0 +1,195 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include "oneapi/dal/table/common.hpp"
+#include "oneapi/dal/table/backend/csr_kernels.hpp"
+#include "oneapi/dal/table/detail/sparse_access_iface.hpp"
+
+namespace oneapi::dal::backend {
+
+class csr_table_impl : public detail::table_template<detail::csr_table_iface, csr_table_impl> {
+public:
+    csr_table_impl()
+            : col_count_(0),
+              row_count_(0),
+              layout_(data_layout::row_major),
+              csr_indexing_(detail::csr_indexing::one_based) {}
+
+    csr_table_impl(std::int64_t column_count,
+                   std::int64_t row_count,
+                   const array<byte_t>& data,
+                   const array<std::int64_t>& column_indices,
+                   const array<std::int64_t>& row_indices,
+                   data_type dtype,
+                   detail::csr_indexing indexing)
+            : meta_(create_csr_metadata(column_count, dtype)),
+              //meta_(create_csr_metadata(1, dtype)),
+              data_(data),
+              column_indices_(column_indices),
+              row_indices_(row_indices),
+              col_count_(column_count),
+              row_count_(row_count),
+              layout_(data_layout::row_major),
+              csr_indexing_(indexing) {
+        using error_msg = dal::detail::error_messages;
+
+        if (row_count <= 0) {
+            throw dal::domain_error(error_msg::rc_leq_zero());
+        }
+
+        if (column_count <= 0) {
+            throw dal::domain_error(error_msg::cc_leq_zero());
+        }
+
+        const int64_t element_count = row_indices_[row_count] - 1;
+        const int64_t dtype_size = detail::get_data_type_size(dtype);
+
+        detail::check_mul_overflow(element_count, dtype_size);
+        if (data.get_count() != element_count * dtype_size) {
```

Comment:
What about checks on indices and indexing?

Reply:
The check of indexing was added/ My suggestion is not repeated full check of indices here, full check were performed in constructor of csr_table. Now we haven't got builder for csr table and check in constructor is enough.